File: test_connection.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/sts"
)

type testConnRequest struct {
	Platform string `json:"platform"`

	AccessKeyID     string `json:"accessKeyId"`
	SecretAccessKey string `json:"secretAccessKey"`
	SessionToken    string `json:"sessionToken"`
	Region          string `json:"region"`

	SubscriptionID string `json:"subscriptionId"`
	TenantID       string `json:"tenantId"`
	ClientID       string `json:"clientId"`
	ClientSecret   string `json:"clientSecret"`

	APIKey string `json:"apiKey"`

	GcpProjectID          string `json:"gcpProjectId"`
	GcpServiceAccountJSON string `json:"gcpServiceAccountJson"`
	ServiceInstanceID     string `json:"serviceInstanceID"`

	BmcUser     string `json:"bmcUser"`
	BmcPassword string `json:"bmcPassword"`
	APIVIP      string `json:"apiVIP"`
	IngressVIP  string `json:"ingressVIP"`
}

type testConnResponse struct {
	OK      bool   `json:"ok"`
	Message string `json:"message"`
}

func handleTestConnection(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		_ = json.NewEncoder(w).Encode(testConnResponse{OK: false, Message: "method not allowed"})
		return
	}
	body, err := io.ReadAll(io.LimitReader(r.Body, 512*1024))
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		_ = json.NewEncoder(w).Encode(testConnResponse{OK: false, Message: "invalid body"})
		return
	}
	var req testConnRequest
	if err := json.Unmarshal(body, &req); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		_ = json.NewEncoder(w).Encode(testConnResponse{OK: false, Message: "invalid json"})
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), 30*time.Second)
	defer cancel()

	var res testConnResponse
	switch strings.TrimSpace(req.Platform) {
	case "aws":
		res = testAWS(ctx, req)
	case "azure":
		res = testAzure(ctx, req)
	case "ibmcloud":
		res = testIBM(ctx, req)
	case "gcp":
		res = testGCP(req)
	case "powervs":
		res = testPowervs(ctx, req)
	case "baremetal":
		res = testBareMetal(req)
	default:
		res = testConnResponse{OK: false, Message: "unknown platform"}
	}
	if !res.OK {
		w.WriteHeader(http.StatusUnprocessableEntity)
	}
	_ = json.NewEncoder(w).Encode(res)
}

func testAWS(ctx context.Context, req testConnRequest) testConnResponse {
	if strings.TrimSpace(req.AccessKeyID) == "" || req.SecretAccessKey == "" {
		return testConnResponse{OK: false, Message: "missing_access_key_or_secret"}
	}
	region := strings.TrimSpace(req.Region)
	if region == "" {
		region = "eu-central-1"
	}
	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithRegion(region),
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			strings.TrimSpace(req.AccessKeyID),
			req.SecretAccessKey,
			strings.TrimSpace(req.SessionToken),
		)),
	)
	if err != nil {
		return testConnResponse{OK: false, Message: "aws_config: " + err.Error()}
	}
	client := sts.NewFromConfig(cfg)
	out, err := client.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
	if err != nil {
		return testConnResponse{OK: false, Message: "aws_sts: " + err.Error()}
	}
	arn := aws.ToString(out.Arn)
	if arn == "" {
		arn = aws.ToString(out.UserId)
	}
	return testConnResponse{OK: true, Message: arn}
}

func testAzure(ctx context.Context, req testConnRequest) testConnResponse {
	tid := strings.TrimSpace(req.TenantID)
	cid := strings.TrimSpace(req.ClientID)
	sec := req.ClientSecret
	if tid == "" || cid == "" || sec == "" {
		return testConnResponse{OK: false, Message: "missing_azure_credentials"}
	}
	tokenURL := fmt.Sprintf("https://login.microsoftonline.com/%s/oauth2/v2.0/token", url.PathEscape(tid))
	form := url.Values{}
	form.Set("client_id", cid)
	form.Set("client_secret", sec)
	form.Set("scope", "https://management.azure.com/.default")
	form.Set("grant_type", "client_credentials")

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return testConnResponse{OK: false, Message: err.Error()}
	}
	httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return testConnResponse{OK: false, Message: "azure_http: " + err.Error()}
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(io.LimitReader(resp.Body, 64*1024))
	if resp.StatusCode != http.StatusOK {
		return testConnResponse{OK: false, Message: fmt.Sprintf("azure_token_%d: %s", resp.StatusCode, truncateMsg(string(b)))}
	}
	var tok struct {
		TokenType string `json:"token_type"`
		ExpiresIn int    `json:"expires_in"`
	}
	if err := json.Unmarshal(b, &tok); err != nil || tok.TokenType == "" {
		return testConnResponse{OK: false, Message: "azure_token_parse"}
	}
	return testConnResponse{OK: true, Message: "azure_ok"}
}

func testIBM(ctx context.Context, req testConnRequest) testConnResponse {
	key := strings.TrimSpace(req.APIKey)
	if key == "" {
		return testConnResponse{OK: false, Message: "missing_ibm_api_key"}
	}
	form := url.Values{}
	form.Set("grant_type", "urn:ibm:params:oauth:grant-type:apikey")
	form.Set("apikey", key)

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://iam.cloud.ibm.com/identity/token", strings.NewReader(form.Encode()))
	if err != nil {
		return testConnResponse{OK: false, Message: err.Error()}
	}
	httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	httpReq.Header.Set("Accept", "application/json")

	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return testConnResponse{OK: false, Message: "ibm_http: " + err.Error()}
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(io.LimitReader(resp.Body, 64*1024))
	if resp.StatusCode != http.StatusOK {
		return testConnResponse{OK: false, Message: fmt.Sprintf("ibm_token_%d: %s", resp.StatusCode, truncateMsg(string(b)))}
	}
	var tok struct {
		TokenType string `json:"token_type"`
	}
	if err := json.Unmarshal(b, &tok); err != nil || tok.TokenType == "" {
		return testConnResponse{OK: false, Message: "ibm_token_parse"}
	}
	return testConnResponse{OK: true, Message: "ibm_ok"}
}

func testGCP(req testConnRequest) testConnResponse {
	if strings.TrimSpace(req.GcpProjectID) == "" {
		return testConnResponse{OK: false, Message: "missing_gcp_project"}
	}
	raw := strings.TrimSpace(req.GcpServiceAccountJSON)
	if raw == "" {
		return testConnResponse{OK: false, Message: "missing_gcp_sa"}
	}
	var js json.RawMessage
	if err := json.Unmarshal([]byte(raw), &js); err != nil {
		return testConnResponse{OK: false, Message: "invalid_gcp_json"}
	}
	return testConnResponse{OK: true, Message: "gcp_ok"}
}

func testPowervs(ctx context.Context, req testConnRequest) testConnResponse {
	if strings.TrimSpace(req.ServiceInstanceID) == "" {
		return testConnResponse{OK: false, Message: "missing_powervs_instance"}
	}
	r := testIBM(ctx, req)
	if !r.OK {
		return r
	}
	return testConnResponse{OK: true, Message: "powervs_ok"}
}

func testBareMetal(req testConnRequest) testConnResponse {
	if strings.TrimSpace(req.BmcUser) == "" || req.BmcPassword == "" {
		return testConnResponse{OK: false, Message: "missing_bmc_credentials"}
	}
	if v := strings.TrimSpace(req.APIVIP); v != "" && !isLikelyIPv4(v) {
		return testConnResponse{OK: false, Message: "invalid_api_vip"}
	}
	if v := strings.TrimSpace(req.IngressVIP); v != "" && !isLikelyIPv4(v) {
		return testConnResponse{OK: false, Message: "invalid_ingress_vip"}
	}
	return testConnResponse{OK: true, Message: "baremetal_ok"}
}

func isLikelyIPv4(s string) bool {
	parts := strings.Split(s, ".")
	if len(parts) != 4 {
		return false
	}
	for _, p := range parts {
		if len(p) == 0 || len(p) > 3 {
			return false
		}
		var n int
		for _, c := range p {
			if c < '0' || c > '9' {
				return false
			}
			n = n*10 + int(c-'0')
		}
		if n > 255 {
			return false
		}
	}
	return true
}

func truncateMsg(s string, max ...int) string {
	m := 200
	if len(max) > 0 {
		m = max[0]
	}
	s = strings.TrimSpace(s)
	if len(s) <= m {
		return s
	}
	return s[:m] + "…"
}

File: api/test-connection.php
<?php
/**
 * POST JSON: same contract as Go /api/test-connection (PHP dev server / Apache).
 */
declare(strict_types=1);

header('Content-Type: application/json; charset=utf-8');
header('X-Content-Type-Options: nosniff');

if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['ok' => false, 'message' => 'method not allowed']);
    exit;
}

$raw = file_get_contents('php://input');
if ($raw === false || strlen($raw) > 524288) {
    http_response_code(400);
    echo json_encode(['ok' => false, 'message' => 'invalid body']);
    exit;
}

$req = json_decode($raw, true);
if (!is_array($req)) {
    http_response_code(400);
    echo json_encode(['ok' => false, 'message' => 'invalid json']);
    exit;
}

$platform = $req['platform'] ?? '';
$result = null;

switch ($platform) {
    case 'aws':
        $result = testAws($req);
        break;
    case 'azure':
        $result = testAzure($req);
        break;
    case 'ibmcloud':
        $result = testIbm($req);
        break;
    case 'gcp':
        $result = testGcp($req);
        break;
    case 'powervs':
        $result = testPowervs($req);
        break;
    case 'baremetal':
        $result = testBareMetal($req);
        break;
    default:
        http_response_code(422);
        echo json_encode(['ok' => false, 'message' => 'unknown platform']);
        exit;
}

if ($result !== null && empty($result['ok'])) {
    http_response_code(422);
}
echo json_encode($result);

function testAws(array $req): array
{
    $key = trim((string)($req['accessKeyId'] ?? ''));
    $sec = (string)($req['secretAccessKey'] ?? '');
    if ($key === '' || $sec === '') {
        return ['ok' => false, 'message' => 'missing_access_key_or_secret'];
    }
    $region = trim((string)($req['region'] ?? 'eu-central-1'));
    if ($region === '') {
        $region = 'eu-central-1';
    }
    $token = trim((string)($req['sessionToken'] ?? ''));

    $aws = trim((string)shell_exec('command -v aws 2>/dev/null'));
    if ($aws === '') {
        return ['ok' => false, 'message' => 'aws_cli_missing'];
    }

    putenv('AWS_ACCESS_KEY_ID=' . $key);
    putenv('AWS_SECRET_ACCESS_KEY=' . $sec);
    putenv('AWS_DEFAULT_REGION=' . $region);
    if ($token !== '') {
        putenv('AWS_SESSION_TOKEN=' . $token);
    } else {
        putenv('AWS_SESSION_TOKEN');
    }
    $out = shell_exec(escapeshellcmd($aws) . ' sts get-caller-identity --output json 2>&1');
    putenv('AWS_ACCESS_KEY_ID');
    putenv('AWS_SECRET_ACCESS_KEY');
    putenv('AWS_DEFAULT_REGION');
    putenv('AWS_SESSION_TOKEN');

    if ($out === null || $out === '') {
        return ['ok' => false, 'message' => 'aws_cli_failed'];
    }
    $j = json_decode($out, true);
    if (is_array($j) && isset($j['Arn'])) {
        return ['ok' => true, 'message' => (string)$j['Arn']];
    }
    return ['ok' => false, 'message' => 'aws_sts: ' . substr(trim($out), 0, 400)];
}

function testAzure(array $req): array
{
    $tid = trim((string)($req['tenantId'] ?? ''));
    $cid = trim((string)($req['clientId'] ?? ''));
    $sec = (string)($req['clientSecret'] ?? '');
    if ($tid === '' || $cid === '' || $sec === '') {
        return ['ok' => false, 'message' => 'missing_azure_credentials'];
    }
    $url = 'https://login.microsoftonline.com/' . rawurlencode($tid) . '/oauth2/v2.0/token';
    $body = http_build_query([
        'client_id' => $cid,
        'client_secret' => $sec,
        'scope' => 'https://management.azure.com/.default',
        'grant_type' => 'client_credentials',
    ]);
    $ctx = stream_context_create([
        'http' => [
            'method' => 'POST',
            'header' => "Content-Type: application/x-www-form-urlencoded\r\n",
            'content' => $body,
            'timeout' => 30,
        ],
    ]);
    $res = @file_get_contents($url, false, $ctx);
    if ($res === false) {
        return ['ok' => false, 'message' => 'azure_http_failed'];
    }
    $j = json_decode($res, true);
    if (is_array($j) && !empty($j['access_token'])) {
        return ['ok' => true, 'message' => 'azure_ok'];
    }
    return ['ok' => false, 'message' => 'azure_token: ' . substr($res, 0, 300)];
}

function testIbm(array $req): array
{
    $key = trim((string)($req['apiKey'] ?? ''));
    if ($key === '') {
        return ['ok' => false, 'message' => 'missing_ibm_api_key'];
    }
    $body = http_build_query([
        'grant_type' => 'urn:ibm:params:oauth:grant-type:apikey',
        'apikey' => $key,
    ]);
    $ctx = stream_context_create([
        'http' => [
            'method' => 'POST',
            'header' => "Content-Type: application/x-www-form-urlencoded\r\nAccept: application/json\r\n",
            'content' => $body,
            'timeout' => 30,
        ],
    ]);
    $res = @file_get_contents('https://iam.cloud.ibm.com/identity/token', false, $ctx);
    if ($res === false) {
        return ['ok' => false, 'message' => 'ibm_http_failed'];
    }
    $j = json_decode($res, true);
    if (is_array($j) && !empty($j['access_token'])) {
        return ['ok' => true, 'message' => 'ibm_ok'];
    }
    return ['ok' => false, 'message' => 'ibm_token: ' . substr($res, 0, 300)];
}

function testGcp(array $req): array
{
    $proj = trim((string)($req['gcpProjectId'] ?? ''));
    if ($proj === '') {
        return ['ok' => false, 'message' => 'missing_gcp_project'];
    }
    $raw = trim((string)($req['gcpServiceAccountJson'] ?? ''));
    if ($raw === '') {
        return ['ok' => false, 'message' => 'missing_gcp_sa'];
    }
    json_decode($raw);
    if (json_last_error() !== JSON_ERROR_NONE) {
        return ['ok' => false, 'message' => 'invalid_gcp_json'];
    }
    return ['ok' => true, 'message' => 'gcp_ok'];
}

function testPowervs(array $req): array
{
    $inst = trim((string)($req['serviceInstanceID'] ?? ''));
    if ($inst === '') {
        return ['ok' => false, 'message' => 'missing_powervs_instance'];
    }
    $r = testIbm($req);
    if (empty($r['ok'])) {
        return $r;
    }
    return ['ok' => true, 'message' => 'powervs_ok'];
}

function testBareMetal(array $req): array
{
    $user = trim((string)($req['bmcUser'] ?? ''));
    $pass = (string)($req['bmcPassword'] ?? '');
    if ($user === '' || $pass === '') {
        return ['ok' => false, 'message' => 'missing_bmc_credentials'];
    }
    $api = trim((string)($req['apiVIP'] ?? ''));
    $ing = trim((string)($req['ingressVIP'] ?? ''));
    if ($api !== '' && !isLikelyIpv4($api)) {
        return ['ok' => false, 'message' => 'invalid_api_vip'];
    }
    if ($ing !== '' && !isLikelyIpv4($ing)) {
        return ['ok' => false, 'message' => 'invalid_ingress_vip'];
    }
    return ['ok' => true, 'message' => 'baremetal_ok'];
}

function isLikelyIpv4(string $s): bool
{
    $parts = explode('.', $s);
    if (count($parts) !== 4) {
        return false;
    }
    foreach ($parts as $p) {
        if ($p === '' || strlen($p) > 3) {
            return false;
        }
        if (!ctype_digit($p)) {
            return false;
        }
        $n = (int)$p;
        if ($n > 255) {
            return false;
        }
    }
    return true;
}
